test/scripts/products/h24-verify-selfmanaged-apicast-and-custom-policy/h24-verify-selfmanaged-apicast-and-custom-policy.go

```diff
@@ -0,0 +1,576 @@
+	message += "b. In the Products section, click API → Integration → Settings → APIcast Self Managed.\n"
+	message += "c. Change the Staging Public Base URL. Replace api-3scale-apicast- with selfmanaged-. \n"
+	message += "d. Click Update Product. \n"
+	message += "e. Click API → Configuration → Promote to Staging \n"
+	message += "f. Copy user_key value (from Staging APIcast - Example curl for testing) to command prompt \n"
+	log.Info(message)
+	reader := bufio.NewReader(os.Stdin)
+	fmt.Print("Waiting for manual step completion. Copy user_key value here and Press enter when done :")
+	userKey, err := reader.ReadString('\n')
+	if err != nil {
+		log.Error("Error: ", err)
+		return "", err
+	}
+	if len(userKey) > 0 {
+		userKey = userKey[:len(userKey)-1]
+	}
+	log.Info("user_key: " + userKey)
+	log.Info("Promote - completed, continue ...")
+	return userKey, err
+}
+
+func promoteSelfManagedAPIcast(ctx context.Context, client k8sclient.Client, promoteManually bool) (string, error) {
+	userKey := ""
+	var err error
+	if promoteManually {
+		userKey, err = promoteSelfManagedAPIcastInUI()
+		if err != nil {
+			log.Error("Error promote Self-managed APIcast. Exit...", err)
+			return "", err
+		}
+	} else {
+		//TODO - WIP - for full automation
+		deleteManagedApicastRoutes(ctx, client)
+		userKey, err = getUserKey()
+		if err != nil {
+			log.Error("Error: ", err)
+			return "TODO. Not implemented. Please use option \"promote-manually=true\" ", err
+		}
+	}
+	return userKey, err
+}
+
+func getUserKey() (string, error) {
+	return "TODO", nil
+}
+
+func validateDeploymentRequest(userKey, routeHost string) (int, error) {
+	httpRequest := "https://" + routeHost + "/?user_key=" + userKey
+	log.Info("HTTP Request: " + httpRequest)
+	resp, err := http.Get(httpRequest)
+	if err != nil {
+		log.Error("HTTP Get error", err)
+		return 0, err
+	}
+	log.Info("Response Code: " + strconv.Itoa(resp.StatusCode))
+	if resp.StatusCode == http.StatusOK {
+		defer resp.Body.Close()
+		bytes, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			log.Error("unable to read response body: ", err)
+			return 0, nil
+		}
+		log.Info("Response Body: " + string(bytes))
+		return http.StatusOK, nil
+	}
+
+	return resp.StatusCode, fmt.Errorf("expected status %v but got %v", http.StatusOK, resp.StatusCode)
+}
+
+func validateDeployment(userKey, routeHost string) (bool, error) {
+	responseCode := 0
+	err := retry(10, 20*time.Second, func() (err error) {
+		responseCode, err = validateDeploymentRequest(userKey, routeHost)
+		return
+	})
+	if err != nil && responseCode != http.StatusOK {
+		return false, err
+	}
+	return true, nil
+}
+
+func retry(attempts int, sleep time.Duration, f func() error) (err error) {
+	for i := 0; ; i++ {
+		err = f()
+		if err == nil {
+			return
+		}
+		if i >= (attempts - 1) {
+			break
+		}
+		time.Sleep(sleep)
+		log.Info("retrying ...")
+	}
+	return fmt.Errorf("after %d attempts, last error: %s", attempts, err)
+}
```

Comment:
I haven't seen this format before. Is it an option to use gos wait.Poll function instead?

Reply:
This function looks fine for me, but yes, suppose we can use  wait.Poll(); will do in https://issues.redhat.com/browse/MGDAPI-3037